core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/CompressRequestStage.java

```diff
@@ -56,23 +61,27 @@ public SdkHttpFullRequest.Builder execute(SdkHttpFullRequest.Builder input, Requ
 
         Compressor compressor = resolveCompressionType(context.executionAttributes());
 
-        // non-streaming
-        if (!isStreaming(context)) {
-            ContentStreamProvider wrappedProvider = input.contentStreamProvider();
-            ContentStreamProvider compressedStreamProvider = () -> compressor.compress(wrappedProvider.newStream());
-            input.contentStreamProvider(compressedStreamProvider);
-            input = updateContentEncodingHeader(input, compressor);
-            return updateContentLengthHeader(input);
+        // non-streaming OR transfer-encoding chunked
+        if (!isStreaming(context) || isTransferEncodingChunked(input)) {
+            compressEntirePayload(input, compressor);
+            updateContentEncodingHeader(input, compressor);
+            if (!isStreaming(context)) {
+                return updateContentLengthHeader(input);
+            } else {
+                return input.removeHeader("Content-Length");
+            }
         }
 
-        /*if (context.requestProvider() == null) {
+        if (context.requestProvider() == null) {
             // sync streaming
-
-        } else {
+            input.contentStreamProvider(new CompressionContentStreamProvider(input.contentStreamProvider(), compressor));
+        }
+        /*else {
```

Comment:
Could you please add a //TODO : here so we wont miss it .

Reply:
added TODO, already have separate branch and implementation for async